### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by olakalisz**
Here is my solution to assignment 3. I am not entirely sure it is optimal but I think it is quite good. Although worst case running time is exponential I think the average is a lot better because I use early termination in my search with isPrefix(string) method. My approach to this problem is to use a dfs-like search on a graph corresponding to a character grid but the search finds all paths from each vertex. It is similar to creating a huge search tree and doing a normal DFS on it with the isPrefix() optimization.  (The tree is formed like, say in given example:
search_tree )
I am very looking forward to hearing what you think of my solution.

### CODE REVIEW

File: Assignment3/Q1.py
Comment (ellapetter) [inline Assignment3/Q1.py]: I think this comment is not needed :)

File: Assignment3/Q1.py
Comment (ellapetter) [inline Assignment3/Q1.py]: I would suggest also adding an example of the a  final graph created

**Review by ellapetter — COMMENTED** [review]
Hi :)
I think your solution is very good!
Only added a couple of comments regarding documentation, but nothing serious.

Have a great weekend!

File: Assignment3/Dictionary.py
Comment (flerdacodeu) [inline Assignment3/Dictionary.py]: For extra credit, you might want to consider an implementation that does not take a prefix list but also a word list.

File: Assignment3/Dictionary.py
Comment (flerdacodeu) [inline Assignment3/Dictionary.py]: Since you change things to uppercase in the `__init__` method, I would recommend you make sure your dictionary works even you are given a non-uppercase string.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py:2]: Actually, in a graph with V vertices and E edges, DFS is O(V+E).
In this case, if you have an NxN grid, what is the worst case complexity? Is it actually exponential?

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: You can use the method `append()` of a list to modify this in place.
```python
>>> l = []
>>> l.append('x')
>>> l
['x']
```

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: Optional suggestion:

It might slightly more readable if you used `[-1, 0, 1]` instead of something that depend on `i` and `j`.
Because you can then write:
```python
  for d_i in [-1, 0, 1]:
    for d_j in [-1, 0, 1]:
       [...]
```
and the rest of your code might also be easier to read (after adjusting it).
Let me know what you think.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: If you get `grid`, you do not actually need to pass also `rows` and `columns` because you can get them using `len()`.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: This is indeed a `key` from the point of view of the `dict`, but what does it represent in your algorithm?

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: As an optimization, you can work throughout with a `set` instead of a `list` so that, if there are many duplicates, you do not build a very long list, just to drop the duplicates later.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: Whenever you have a dict, you can make it clearer by specifying what the keys and the values are. E.g.:

```python
  a dictionary, mapping integers into their string representations.
```

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: You might want to call this `a pair of integers`.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: Usually, if you accumulate the `words` in an argument, you might want to also return it. In this case, above you assign `words` to the result of `word_search_recursive`, but they actually should contain the same identical list even without assigning.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: As noted elsewhere, we now have two separate pieces of code (the implementation of this method and of the dictionary) which both need to know that the convention is to convert everything to upper case.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: Is `edge` actually an edge here?

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: Optimization (if you feel up for it):
Right now you build and store the entire graph, but there might be many cases that are actually never explored. The grid, in a way, is a more compact representation of the same graph. Can we visit the graph without actually needing to store it? Happy to discuss some ideas to point you in the right direction if interested.

**Review by flerdacodeu — CHANGES_REQUESTED** [review]
Well done! A few suggestions for you but this looks nice.

**Comment by olakalisz** [comment]
Hi!
Thank you both so much for reviewing my code. I have changed it slightly now. The main idea is still the same but I am no longer using the graph corresponding to a grid but I perform my search directly on a grid.
When you find some time I would really appreciate if you could tell me what you think about my changes.

File: Assignment3/Dictionary.py
Comment (olakalisz) [reply Assignment3/Dictionary.py]: I have changed it now. __init__ function creates the prefix_list array from the word_list array.

File: Assignment3/Dictionary.py
Comment (olakalisz) [reply Assignment3/Dictionary.py]: It checks for the uppercase word in a dictionary now.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py:2]: Ooops! I was overthinking it. In case of NxN grid the running time is O(N^4) (N^2 calls of DFS which takes O(N^2) time N^2 vertices and at most 8N^2 edges, actually because each edge connectes two grid entried it might be even bounded by 4N^2 edges). In general case, I guess, it's O((MxN)^2) M-number of rows and N columns in a grid.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: That's a really good point! I have changed it now, no longer using the graph. Looking forward to hearing what you think of this solution.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I am not using this method anymore but I'll bear it in mind the next time!

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: That's a really good point, it would make it a lot clearer! I have changed my implementation now and I'm not creating the graph at all but I'll definately add an example of a data-structure that I'm creating in my future solutions!

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I did something similar to what you suggested but I just have one for loop that does 8 iterations (8 adjacent entries). Please tell me what you think about this solution.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I used this method in my new implementation of Dictionary 👍

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: True, it was not very useful 😺

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: Ohh, I thought we should pass them as an argument as it was mentioned in an assignment question, changed now!

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: It was either a vertex or simply an grid entry, now I called it 'adjacent'.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: Done! Also changed the implementation so it actually returns a list (with no duplicates) as asked it assignment question.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: Done!

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I have changed it now. But in a word_search function I change all found words to uppercase in order to make testing easries. Dunno if that's okay. Please let me know what you think.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: EDIT: It was just a grid entry and I don't use it anymore. I just use the coordinates instead.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: It's an adjacent entry, I call it 'adjacent' now

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I have changed it, it is a void function now 👍 It was also easier to use the 'visited' 2D array without returning in two different places!

File: Assignment3/Q1.py
Comment (flerdacodeu) [reply Assignment3/Q1.py:2]: Yes, that's right! The size of the graph is quadratic with respect to N, so each DFS is O(N^2), and you need to have separate DFS for each cell in the grid (node in the graph) so you get O(N^4).  As you mentioned, the number of edges is also bound by the number of edges in the graph.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: nit:
weather -> whether
it is still not visited -> it is not yet visited

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: from the 'vertex' -> from the 'start' vertex

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: The description of `word_list` is missing something: as written, it seems that the function is given as an input a words that can be formed by the grid, rather than generating it.

What you want to make clear is:
* the function should add to `word_list` any words it finds
* at the start, `word_list` may contain words words we have already found

In the above description of the function, you might also mention the role that `word_list` plays.

At first, a user of this function might assume this function returns the list.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: Usually the "Returns" section is left out if the function does not return a value. But what you wrote here is relevant! As mentioned above this sentence could go in the main description and the description of the `word_list` argument instead.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: Since you later (in the outer function) convert words to upper case, this check for duplicates is not enough. Two options:
* convert to upper case here before checking whether it is `word_list`
* when you convert to upper case in the other function, also check for duplicates

Nice extra: you can add a test that verifies this behavior, e.g.:
C a T
z c t
contains: `CaT`, `Cat`, `caT`, and `cat`, but your list should contain in the end only `CAT`.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: These are essentially constants. At the moment they are created each time we hit this and therefore O(N^4) times.

You can instead define top-level constants, e.g.:
```python
# Two arrays to keep track of adjacent entries.
_ROW_ADJACENT = [-1, -1, -1, 0, 0, 1, 1, 1]
_COLUMNS_ADJACENT = [-1, 0, 1, -1, 1, -1, 0, 1]
```
The name started with `_` because these are meant to be private to this module.
You can add your comment on top of them.

As a plus, you can define a single constant containing pairs of row and column deltas instead: this means that each pair is clearly identified (right now, you need to understand both variables together).

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: Instead of `range(8)`, you can refer directly to the variables with `range(len(row_adjacent)`. Using a single variable storing pairs (as mentioned above) would give you the extra benefit of:
* you do not need to make sure you have the two lists having the same length
* you can more easily iterate with something like:
```python
  for (row_delta, column_delta) in adjacent_deltas:
    [...]
```

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: Using a `set` for the final result is actually not a bad idea.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: When using a keyword argument (the optional argument `prefix` in this case), it is a useful convention to always mention it by name, i.e., in this case:
```python
    word_search_recursive(dictionary, grid, visited, adjacent, word_list, prefix=prefix)
```
What you did is not incorrect, but using the explicit name helps if you add a second optional argument (or change the order of optional arguments), as things will always work correctly with the named argument. The current implementation depends on the fact that prefix is the first optional argument.

File: Assignment3/Q1_test.py
Comment (flerdacodeu) [inline Assignment3/Q1_test.py]: Optional suggestion about testing and constants.

The current tests require a bit of work to figure out what exactly is being tested because there are a number of variables being set in the `setUp()` method and then used in tests. The main concern is that if you look at one of the tests:
```python
  for word in self._given_desired_output:
    self.assertTrue(word in Q1.word_search(self._given_grid, self._given_dictionary))
```
it is not immediately obvious what is going on.

Consider, for comparison:
```python
  self.assertIn('CAT', Q1.word_search(_grid("CAD|TGO"), _dictionary('cat', 'dog'))
```
But how can we get this to work?

We can start by defining helper functions `_grid` and `_dictionary`:
```python
def _grid(string):
  """Creates a 2D array of characters from a string.

  The string represents the grid by having the characters in each row separated by a '|'.

  Args:
    string: a string, representation of the grid as a single string.

  Returns:
    a 2-D array of characters corresponding to the grid given.
  """
  grid = []
  for row in string.split('|'):
    grid.append([ch for ch in row])
    # Make sure each row has the same length.
    if len(grid[0]) != len(grid[-1]):
      raise ValueError("not all rows have the same length")
  return grid

def _dictionary(*words):
  """Create a dictionary with the given words.

  Args:
    *words: a list of strings, the words to be added to the dictionary.

  Returns:
    an instance of Q1.Dictionary containing the given words.
  """
  return Q1.Dictionary(words)
```
with these as helper functions defined in the test module, we can make the tests themselves more compact, without having to copy and paste too much.

It is sometimes helpful to have the `setUp()` create things for us, but that's mostly the case when things are shared by all the tests, not just some of them, like in this case.

File: Assignment3/Dictionary.py
Comment (flerdacodeu) [reply Assignment3/Dictionary.py]: Yes, this looks nice and makes it a lot easier (and less error prone) to create a dictionary.

File: Assignment3/Q1.py
Comment (flerdacodeu) [reply Assignment3/Q1.py]: The assignment problem did explicitly mention these, but it turns out they are not actually needed, so it is nicer to remove them.

I think the reason the assignment is phrased that way is probably because the person who wrote thought of a particular implementation, where this would actually be a requirement (think if the input did pass a single string as the representation of the grid, then we would need to know how to break it apart into rows and columns) or maybe though of how the implementation would for in C++ (where the arrays do not actually have a size defined).

Either way, I felt this would be a good example of how requirements might change when one understands the problem better and we should know that, unlike a homework assignment where one has to exactly match the question, in the real world changing the question while solving the problem is actually the best way forward in a case like this.

File: Assignment3/Q1.py
Comment (flerdacodeu) [reply Assignment3/Q1.py]: The assignment question does say `set` rather than `list` I believe.
What you did is actually now solving the problem of duplicates (but, as mentioned elsewhere, it has a problem with case).
What I meant, however, was that if `word_list` is instead `word_set`, you can just call `word_set.add()` without having to worry if it in the set or not, because `set` takes care of that:
```python
>>> word_set = set()
>>> word_set.add('cat')
>>> word_set
set(['cat'])
>>> word_set.add('cat')
>>> word_set
set(['cat'])
>>>
```

**Review by flerdacodeu — CHANGES_REQUESTED** [review]
Really nice. Thanks for all the effort you put in it!

I added a few more suggestions. Only real concern is the fact that you might have duplicates with different combinations of lower-case and upper-case characters. But it should a simple fix.

One suggestion: add a test first that fails because of the problem above, and then fix the code and see the test is passing. This is called TDD (test-driven development) and it is often a useful way to make sure you write better tests and code.

File: Assignment3/Dictionary.py
Comment (flerdacodeu) [inline Assignment3/Dictionary.py]: One additional comment I forgot!

Consider a dictionary with `dog`, `cat`, and `caterpillar`.

Currently it will return:
```python
>>> d = Dictionary(['dog', 'cat', 'caterpillar'])
>>> d.isPrefix('dog')
True
>>> d.isWord('dog')
True
>>> d.isPrefix('cat')
True
>>> d.isWord('cat')
True
>>>
```
However, `cat` can be extended to `caterpillar` but `dog` cannot be extended to anything. So, once you get to `dog` you could stop searching, but your implementation will actually try to find words that start with `dog` plus on more character, even if there not any.

Comment (olakalisz) [reply Assignment3/Dictionary.py]: It is changed now. So for the final word (say 'dog') isPrefix returns False. And in the code later I first check isWord() and then continue searching if isPrefix() returns True. I hope that's the change you were thinkg of 👍 I might have included it as an optimization earlier but I thought isPrefix() is supposed to be defined exactly in a way that was described in an assignment question.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: Done!

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I changed it slightly and I'm using a set instead of the list the whole way thourgh.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: Done!

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I've changed the description, please let me know if it's clearer now.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I don't mention return value in this docstring anymore.

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: So I changed my implementation a bit and I use the set data strusture from the very beginning. Thus it should not be an issue anymore 👍 (I also convert aa string to upper-case in if condition now although I guess as I'm using set the second part in if condition here
```if dictionary.isWord(prefix) and prefix.upper() not in found_words:```
is not really needed)

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I did what you suggested using one table with pair of coordinates. This was a very useful comment, thank you!

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: Done!

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I've changed it! Thank you for the comment, haven't known about keyword arguments before.

File: Assignment3/Q1_test.py
Comment (olakalisz) [reply Assignment3/Q1_test.py]: I did what you suggested here. I just couldn't get the ```*words``` argument to work, I think it was creating a tuple not a list and then word_list[i].upper() (in Dictionary) was giving me an error. So I am passing a list ```words``` as argument now. Couldn't figure out how to fix ```*words``` version.

File: Assignment3/Dictionary.py
Comment (flerdacodeu) [inline Assignment3/Dictionary.py]: I believe this now adds the empty string, which is unnecessary.
You can use:
```python
  for i in range(1, len(word)::
```
do skip the value `0`.

File: Assignment3/Dictionary.py
Comment (flerdacodeu) [inline Assignment3/Dictionary.py]: It would be better for `self.__word_list` and `self.__prefix_list` to actually be sets. This would make look-up by name faster, which is the main operation you do on them.

File: Assignment3/Q1_test.py
Comment (flerdacodeu) [reply Assignment3/Q1_test.py]: Sorry. I did not actually test it :)
If you want to fix that issue, you can use the `list()` function:
```
>>> x = ('a', 'b')
>>> x
('a', 'b')
>>> list(x)
['a', 'b']
>>>
```

File: Assignment3/Q1.py
Comment (flerdacodeu) [reply Assignment3/Q1.py]: Correct. You might want to remove the `prefix.upper() not in found_words' condition for the set, since adding to the set does not add it twice.

File: Assignment3/Q1.py
Comment (flerdacodeu) [reply Assignment3/Q1.py]: Looks good.

File: Assignment3/Q1.py
Comment (flerdacodeu) [inline Assignment3/Q1.py]: The phrase `there are no duplicated words in a set` is a bit redundant since a `set` cannot contain duplicates.

**Review by flerdacodeu — APPROVED** [review]
Thanks!

File: Assignment3/Dictionary.py
Comment (olakalisz) [reply Assignment3/Dictionary.py]: Done!

File: Assignment3/Q1.py
Comment (olakalisz) [reply Assignment3/Q1.py]: I removed that 👍

File: Assignment3/Dictionary.py
Comment (olakalisz) [reply Assignment3/Dictionary.py]: Changed

**Comment by olakalisz** [comment]
Thank you so much for your review! I think I'm quite happy with it now, I'll merge it 👍

**Comment by ellapetter** [comment]
@pistacja96 
Sorry, I didn't have time to look at it again before you merged it..
I gave it a look now and it looks great in my opinion :)

**Pull request merged**
